### PR TITLE
[Nano] How-To Guides: Accelerate PyTorch Training with IPEX & Multi-instance & BF16 & Channels last

### PR DESCRIPTION
**Pull request by Siritao**
### 1. Why the change?

Add hands-on guides for Nano users to apply __IPEX / Multi-instance / BF16 / Channels last__ to their PyTorch training. How-to guides aim at giving multiple bite-sized, task-oriented, and executable examples for user to check when they need to do similar tasks.

### 2. User API changes

N/A

### 3. Summary of the change 

Including __4__ how-to guides, each provides __2__ approaches: subclass `TorchNano` / use `@nano` decorator.
- [x] [How to accelerate a PyTorch application on training workloads through Intel® Extension for PyTorch*](https://siruitestdoc.readthedocs.io/en/nano-howto-pytorch-training/doc/Nano/Howto/Training/PyTorch/accelerate_pytorch_training_ipex.html)
- [x] [How to accelerate a PyTorch application on training workloads through multiple instances](https://siruitestdoc.readthedocs.io/en/nano-howto-pytorch-training/doc/Nano/Howto/Training/PyTorch/accelerate_pytorch_training_multi_instance.html)
- [x] [How to use the channels last memory format in your PyTorch application for training](https://siruitestdoc.readthedocs.io/en/nano-howto-pytorch-training/doc/Nano/Howto/Training/PyTorch/pytorch_training_channels_last.html)
- [x] [How to conduct BFloat16 Mixed Precision training in your PyTorch application](https://siruitestdoc.readthedocs.io/en/nano-howto-pytorch-training/doc/Nano/Howto/Training/PyTorch/accelerate_pytorch_training_bf16.html)

### 4. How to test?
- [x] Document test: https://siruitestdoc.readthedocs.io/en/nano-howto-pytorch-training/doc/Nano/Howto/index.html#pytorch
- [x] GitHub Notebook Preview: https://github.com/Siritao/BigDL/tree/nano-howto-pytorch-training/python/nano/tutorial/notebook/training/pytorch
- [x] Notebook test locally

### CODE REVIEW

File: python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "> 📝 **Note**\n",
+    ">\n",
+    "> Two more parameters will be helpful if you want to specify the distributed backend and CPU cores for each process.\n",
+    ">\n",
+    "> * distributed_backend – use which backend in distributed mode, defaults to ‘subprocess’, now available backends are ‘spawn’, ‘subprocess’ and ‘ray’\n",
+    "> * cpu_for_each_process – a list of length `num_processes`, in which each item is a list of CPU indices to specify the cores used by each process, if None, CPU cores will be distributed evenly by all processes, only take effect when `num_processes` > 1"
```
Comment (Oscilloscope98) [inline python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb:231]: I think it may be not necessary to copy the api doc definition here for these two parameters. The important thing here is to mention that for `TorchNano`, `spawn` is supported as a distributed backend, but for `@nano` it does not.

You could add reference to the API doc here for more detailed parameter explanation if you want.

Same for the note box for `@nano`

**Review by Oscilloscope98 — COMMENTED** [review]
Others LGTM.

Please rebased on main after the two related how-to guides are merged.

**Comment by Oscilloscope98** [comment]
The related PR regarding converting `TorchNano` and `@nano` is merged. Please rebase and adding related links according to [this](https://github.com/intel-analytics/BigDL/pull/5984#discussion_r1045374045) :)

File: python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb (hunk 1 above)
Comment (Siritao) [reply python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb]: Thanks for your comments! I've modified all of them and rebased my branch.

**Comment by Siritao** [comment]
> 

Thx, all is done.

**Comment by Siritao** [comment]
> The related PR regarding converting `TorchNano` and `@nano` is merged. Please rebase and adding related links according to [this](https://github.com/intel-analytics/BigDL/pull/5984#discussion_r1045374045) :)

Thx, all is done.

File: python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_bf16.ipynb [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+    "\n",
+    "    return train_dataloader"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## A) Subclass TorchNano"
```
Comment (Oscilloscope98) [inline python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_bf16.ipynb:167]: The prefix of subtitle "Subclass TorchNano" and "Use @nano decorator" are not consistent. You are using:
```markdown
## A) Subclass TorchNano

## B. Use `@nano` decorator
```
Please make them consistent :)

That is:

```markdown
## A) Subclass TorchNano

## B) Use `@nano` decorator
```

Same for other notebooks.

File: python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+   "source": [
+    "# Accelerate PyTorch Training using Multiple Instances"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`TorchNano` (`bigdl.nano.pytorch.TorchNano`) supports multi-instance training that can make the most of hardwares with multiple CPU cores or sockets (especially when the number of cores is large). Here we provide __2__ ways to achieve this: A) subclass `TorchNano` or B) use `@nano` decorator. You can choose the appropriate one depending on your (preferred) code structure."
```
Comment (Oscilloscope98) [inline python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb:24]: It seems like that this sentence is not complete:

`TorchNano` (`bigdl.nano.pytorch.TorchNano`) supports multi-instance training that can make the most of hardwares with multiple CPU cores or sockets (especially when the number of cores is large) ...?

Do you mean ".... training that can benefit from the most hardwares ..." ?

File: python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+    "> \n",
+    "> Nano supports multiple distributed backends, i.e., ‘subprocess’ (default), ‘spawn’ and ‘ray’, you can switch according to your preferences by setting `distributed_backend`.\n",
+    "> \n",
+    "> By setting `num_processes`, CPU cores will be automatically and evenly distributed among specific number of processes, to avoid conflicts and maximize training throughput. If you would like to specify the CPU cores used by each process, You could set `cpu_for_each_process` to a list of length `num_processes`, in which each item is a list of CPU indices.\n",
+    "> \n",
+    "> When using data-parallel training, the batch size is equivalent to becoming `num_processes` times larger, so the learning rate warm-up strategy that gradually increase the learning rate to `num_processes` times is needed. Nano enables this practice by default through `auto_lr=True`.\n",
+    "> \n",
+    "> Please refer to the [API doc](https://bigdl.readthedocs.io/en/latest/doc/PythonAPI/Nano/pytorch.html#module-bigdl.nano.pytorch.torch_nano) for more detailed information regarding multi-instance related parameters in `TorchNano`"
```
Comment (Oscilloscope98) [inline python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb:246]: Maybe this part can be adjusted a little bit (feel free to edit it if you find somewhere not reasonable):

---
```markdown
By setting `num_processes`, CPU cores will be automatically and evenly distributed among specific number of processes, to avoid conflicts and maximize training throughput. If you would like to specify the CPU cores used by each process, You could set `cpu_for_each_process` to a list of length `num_processes`, in which each item is a list of CPU indices.

Currently, `‘subprocess’` (default), `‘spawn’` and `‘ray’` are supported as `distributed_backend` for `TorchNano`.

Also note that, when using data-parallel training, the batch size is equivalent to becoming `num_processes` times larger. The learning rate warm-up strategy that gradually increases the learning rate to `num_processes` times is a compensate to achieve the same effect as single instance training. Nano enables this strategy by default through `auto_lr=True`.
```
---

As I think that the above contents are detailed enough. Maybe there is no need to link to the API doc here.

File: python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    "> 📝 **Note**\n",
+    ">\n",
+    "> Three more parameters will be helpful if you want to customize your multi-instance training.\n",
+    "> \n",
+    "> Nano supports multiple distributed backends, i.e., ‘subprocess’ (default), ‘spawn’ (__not supported when using__ `@nano` __decorator, please subclass__ `TorchNano` __to adopt ‘spawn’__) and ‘ray’, you can switch according to your preferences by setting `distributed_backend`.\n",
+    "> \n",
+    "> By setting `num_processes`, CPU cores will be automatically and evenly distributed among specific number of processes, to avoid conflicts and maximize training throughput. If you would like to specify the CPU cores used by each process, You could set `cpu_for_each_process` to a list of length `num_processes`, in which each item is a list of CPU indices.\n",
+    "> \n",
+    "> When using data-parallel training, the batch size is equivalent to becoming `num_processes` times larger, so the learning rate warm-up strategy that gradually increase the learning rate to `num_processes` times is needed. Nano enables this practice by default through `auto_lr=True`."
+   ]
```
Comment (Oscilloscope98) [inline python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb:334]: Maybe this part can be adjusted a little bit (feel free to edit it if you find somewhere not reasonable):

---

```markdown
By setting `num_processes`, CPU cores will be automatically and evenly distributed among specific number of processes, to avoid conflicts and maximize training throughput. If you would like to specify the CPU cores used by each process, You could set `cpu_for_each_process` to a list of length `num_processes`, in which each item is a list of CPU indices.

Currently, `‘subprocess’` (default), and `‘ray’` are supported as `distributed_backend` for `@nano` decorator. (`'spawn'` is not supported by `@nano`)

Also note that, when using data-parallel training, the batch size is equivalent to becoming `num_processes` times larger. The learning rate warm-up strategy that gradually increases the learning rate to `num_processes` times is a compensate to achieve the same effect as single instance training. Nano enables this strategy by default through `auto_lr=True`.
```
---

As I think that the above contents are detailed enough. Maybe there is no need to link to the API doc here.

**Review by Oscilloscope98 — COMMENTED** [review]
Others LGTM.

Please make sure that tthese 4 notebooks can run successfully :)

File: python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_bf16.ipynb [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "> ⚠️ **Warning**\n",
+    "> \n",
+    "> Using BFloat16 precision with `torch < 1.12` may result in extremely slow training."
```
Comment (MeouSker77) [inline python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_bf16.ipynb:91]: running bf16 on CPU without bf16 instruction support is also slow.

**Review by MeouSker77 — APPROVED** [review]
others LGTM

File: python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_ipex.ipynb [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+   "source": [
+    "# Accelerate PyTorch Training using Intel® Extension for PyTorch*"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "[Intel® Extension for PyTorch*](https://github.com/intel/intel-extension-for-pytorch) (also known as IPEX) can boost performance on Intel hardware with up-to-date features optimizations. By using `TorchNano` (`bigdl.nano.pytorch.TorchNano`), you can make very few code changes to accelerate training loops via IPEX. Here we provide __2__ ways to achieve this: A) subclass `TorchNano` or B) use `@nano` decorator. You can choose the appropriate one depending on your (preferred) code structure."
```
Comment (TheaperDeng) [inline python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_ipex.ipynb:24]: "with up-to-date features optimizations." -> with AVX-512 Vector Neural Network Instructions (AVX512 VNNI) and Intel® Advanced Matrix Extensions (Intel® AMX) on Intel CPUs.

File: python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb (hunk 3 above)
Comment (TheaperDeng) [reply python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb:24]: I guess the correct representative might be "that can make **full usage** of hardwires with ..."

**Review by TheaperDeng — COMMENTED** [review]
others LGTM

File: python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb (hunk 3 above)
Comment (Siritao) [reply python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_multi_instance.ipynb:24]: Here I used __make the most of__ as a phrase equivalent to __make full usage of / take full advantage of__. I will replace it to disambiguate.

File: python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_bf16.ipynb (hunk 6 above)
Comment (Oscilloscope98) [reply python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_bf16.ipynb:91]: From my understanding, running bf16 is not supported on CPU without bf16 instruction sets, either with or without IPEX. Maybe we should put this sentence in the warning box here?
image

For me, the sentence here sounds like that bf16 with IPEX can be a work around for CPU without bf16 instruction sets, which seems to be not correct.

Comment (MeouSker77) [reply python/nano/tutorial/notebook/training/pytorch/accelerate_pytorch_training_bf16.ipynb:91]: `bf16` requires bf16 instruction, such as `avx512_bf16`, while `IPEX+bf16` only requires `avx512` instruction (`avx512` doesn't belong to bf16 instruction), so `IPEX+bf16` is indeed a workaround.
